src/raft_server.c

```diff
@@ -933,12 +948,10 @@ int raft_apply_entry(raft_server_t* me_)
             raft_node_set_addition_committed(node, 1);
             break;
         case RAFT_LOGTYPE_DEMOTE_NODE:
-            if (node)
-                raft_node_set_voting_committed(node, 0);
+            raft_node_set_voting_committed(node, 0);
             break;
         case RAFT_LOGTYPE_REMOVE_NODE:
-            if (node)
-                raft_remove_node(me_, node);
+            raft_node_set_addition_committed(node, 0);
```

Comment:
What cases need this and those changes in raft_apply_all? The membership change algorithm allows at most one uncommitted membership change entry, so we can simply remove the node here, no? (I must have missed your considerations.)

Reply:
When we're making live changes - yes!

But if we add a new node and it receives a series of membership change entries as part of "appendentries" from the leader it has to commit and apply all of them locally. The best way I figured out was to keep track of additions and removals and making changing the actual node list at the end (deleting all nodes which have been marked for removal, i.e. addition_committed == false). 

This is a little tricky to understand if we don't see RDB/raft in action. The problem is that if we have log entries affecting nodes ranks 'x', 'y' and 'z' -- we need those nodes to be present in the node list even before those log entries have been applied. 

Let me know if you'd like to discuss this further.

Reply:
I think I've gained a feeling of the issue. The membership change handling in raft_apply_entry appears to cause both the issue you described here and the issue in rdb_raft_start. By the time we apply a membership change entry i in raft_apply_entry, the node in question may have already been changed by a latter entry j when j is offered.